### PR TITLE
[AIRFLOW-XXX] Add a git pre-commit hook template

### PR DESCRIPTION
**Pull request by feluelle**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-XXX
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

File: .githooks/pre-commit [hunk 1]
```diff
@@ -0,0 +1,74 @@
+#!/bin/sh
+
+# This Git Hook is based on the .travis.yml to avoid as many errors in the Travis CI Pipeline as possible.
+# It is optional, but recommended to have at least the pre-tests section enabled.
+
+# Please make sure that you have your development environment setup and ready. See the CONTRIBUTING.md for more information.
+
+# NOTE: You need to set the following variables before running this hook.
+# A python virtual environment is needed for the pre-tests to run in.
+VIRTUAL_ENV_PATH=/Users/feluelle/venv/apache-airflow/bin/activate
```
Comment (BasPH) [inline .githooks/pre-commit:10]: You probably want to remove your own username here

**Review by BasPH — COMMENTED** [review]
I can see the benefit of a pre-commit hook for small things like flake8.

However to me it feels weird having to edit the pre-commit hook for each feature to set the specific test, and requiring running Docker containers with a specific name. At the very least, I'd add some documentation on usage.

In my experience, pre-commit hooks with long running processes is something that annoys people, rather than helping, so I don't think many people would use it in the this way. But feel free to clarify if your experience is different :)

File: .githooks/pre-commit (hunk 1 above)
Comment (feluelle) [reply .githooks/pre-commit]: > However to me it feels weird having to edit the pre-commit hook for each feature to set the specific test, and requiring running Docker containers with a specific name. At the very least, I'd add some documentation on usage.

You can of course just uncomment the `run_tests` section if you dont want it to run before each commit. I personally like it. The whole pipeline lasts like 3 min. I can understand you if you say you want to have a commit quickly done. But for me its the feeling that I am almost certain that the code I commit passes Travis and I dont need to check it every time.
I have two docker containers one for python 3 and one for python 2 that I use to run tests, but if you like I can add some lines to run the tests in the python venv as well or instead.

> In my experience, pre-commit hooks with long running processes is something that annoys people, rather than helping, so I don't think many people would use it in the this way. But feel free to clarify if your experience is different :)

When I implement a new feature at first I edit the pre-commit hook change it once to the name of the new test class I am going to create along with the feature class. I am doing a kinda test-driven approach splitting the screen into one half containing the test class and one containing the feature. I can focus completly on writing logic for the feature and its tests instead of having to think about whether it works for python 2 or 3 or having some kinda linting issues I forget to check, etc. - I am done when the commit comes through. I can checkout to a commit that has a working progress of a feature. It does not need to be finished but it works until this point.



**tl;dr**
I think you can use this script even if you don't use it as git pre-commit. In PyCharm I have a Run Config running this script - I only need to click a button. But in case I forget to click it right before I commit changes there might be an issue in Travis that could probably have been avoided when I had used it as git pre-commit hook.

**Comment by feluelle** [comment]
Thanks @BasPH for the review :)

> However to me it feels weird having to edit the pre-commit hook for each feature to set the specific test, and requiring running Docker containers with a specific name. At the very least, I'd add some documentation on usage.

You can of course just uncomment the `run_tests` section if you dont want it to run before each commit. I personally like it. The whole pipeline lasts like 3 min. I can understand you if you say you want to have a commit quickly done. But for me its the feeling that I am almost certain that the code I commit passes Travis and I dont need to check it every time.
I have two docker containers one for python 3 and one for python 2 that I use to run tests, but if you like I can add some lines to run the tests in the python venv as well or instead.

> In my experience, pre-commit hooks with long running processes is something that annoys people, rather than helping, so I don't think many people would use it in the this way. But feel free to clarify if your experience is different :)

When I implement a new feature at first I edit the pre-commit hook change it once to the name of the new test class I am going to create along with the feature class. I am doing a kinda test-driven approach splitting the screen into one half containing the test class and one containing the feature. I can focus completly on writing logic for the feature and its tests instead of having to think about whether it works for python 2 or 3 or having some kinda linting issues I forget to check, etc. - I am done when the commit comes through. I can checkout to a commit that has a working progress of a feature. It does not need to be finished but it works until this point.



**tl;dr**
I think you can use this script even if you don't use it as git pre-commit. In PyCharm I have a Run Config running this script - I only need to click a button. But in case I forget to click it right before I commit changes there might be an issue in Travis that could probably have been avoided when I had used it as git pre-commit hook.

File: .githooks/pre-commit (hunk 1 above)
Comment (feluelle) [reply .githooks/pre-commit:10]: True. Thank you. Good catch :) I am gonna use `$HOME` instead.

Comment (feluelle) [reply .githooks/pre-commit:10]: ..or probably better to define them not in the script directly and instead `export` these variables before running the script.

**Comment by feluelle** [comment]
@BasPH I think you are right it will be really difficult to have a pre-commit file that is easy configurable for everyone. 

Therefore I would really like to see what other think about this.

**Comment by Fokko** [comment]
I'm on @BasPH. I never use pre-commit hooks because I haven't found them useful. I'm still open to it if someone else finds it useful.

**Comment by feluelle** [comment]
I am closing it for now and maybe bring this topic up in a DISCUSS mail thread next week or so to get more opinions. :)
